### DISCUSSION TITLE
how-to-register-a-service-with-multiple-interfaces-for-in-asp-net-core-di/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#116] [Announcements]
# how-to-register-a-service-with-multiple-interfaces-for-in-asp-net-core-di/

In this post I describe how to register a class that implements multiple interfaces with the ASP.NET Core DI container, and how to avoid a common bug

https://andrewlock.net/how-to-register-a-service-with-multiple-interfaces-for-in-asp-net-core-di/

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at September 22 2018, 01:41, in reply to [Steve Hansen](https://disqus.com/by/xiugraag/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/116#discussioncomment-6046176)</em>

---

<p>That's a good point, as long as IBar implements IFoo of course. Otherwise you'd need to do a direct cast, e.g. </p><pre><code>AddSingleton&lt;ibar&gt;(x =&gt; (IBar)x.GetRequiredService&lt;ifoo&gt;());</code></pre><p></p><p>Also, I don't think performance would be a big deal for singletons, as the resolved instance should be cached for subsequent requests. It could make a difference for scoped instances though.</p>

[a comment by an automated account is not shown]

**Comment by slimshader** [comment]
The problem with suggested solutions is that now Foo is explicitly registered, anybody can inject it which might be not desired (as we only want to expose interfaces through DI)
